pkg/ccl/importccl/import_pgdump.go

```diff
@@ -0,0 +1,503 @@
+				return err
+			}
+		}
+	case *tree.BeginTransaction, *tree.CommitTransaction:
+	case *tree.Insert, *tree.CopyFrom, *tree.Delete, copyData:
+		// handled during the data ingestion pass.
+	case *tree.CreateExtension, *tree.CommentOnDatabase, *tree.CommentOnTable,
+		*tree.CommentOnIndex, *tree.CommentOnConstraint, *tree.CommentOnColumn, *tree.SetVar, *tree.Analyze,
+		*tree.CommentOnSchema:
+		// These are the statements that can be parsed by CRDB but are not
+		// supported, or are not required to be processed, during an IMPORT.
+		// - ignore txns.
+		// - ignore SETs and DMLs.
+		// - ANALYZE is syntactic sugar for CreateStatistics. It can be ignored
+		// because the auto stats stuff will pick up the changes and run if needed.
+		return wrapErrorWithUnsupportedHint(errors.Errorf("unsupported %T statement: %s", stmt, stmt))
+	case *tree.CreateType:
+		return errors.New("IMPORT PGDUMP does not support user defined types; please" +
+			" remove all CREATE TYPE statements and their usages from the dump file")
+	case error:
+		if !errors.Is(stmt, errCopyDone) {
+			return stmt
+		}
+	default:
+		return wrapErrorWithUnsupportedHint(errors.Errorf("unsupported %T statement: %s", stmt, stmt))
+	}
+	return nil
+}
+
+// parseDDLStatementsFromDumpFile parses the DDL statements from the dump file,
+// and replaces all qualified object names to point to the temporary database
+// being imported into.
+// This method returns a postgresDDLHandler that contains the buffered DDL
+// statements along with other relevant metadata.
+func parseDDLStatementsFromDumpFile(
+	ctx context.Context,
+	evalCtx *tree.EvalContext,
+	p sql.JobExecContext,
+	dumpFile string,
+	format roachpb.IOFileFormat,
+	maxRowSize int,
+	parentID descpb.ID,
+) (postgresDDLHandler, error) {
+	handler := postgresDDLHandler{bufferedDDLStmts: make([]string, 0)}
+	// Open the dump file.
+	store, err := p.ExecCfg().DistSQLSrv.ExternalStorageFromURI(ctx, dumpFile, p.User())
+	if err != nil {
+		return handler, err
+	}
+	defer store.Close()
+
+	raw, err := store.ReadFile(ctx, "")
+	if err != nil {
+		return handler, err
+	}
+	defer raw.Close()
+	reader, err := decompressingReader(raw, dumpFile, format.Compression)
+	if err != nil {
+		return handler, err
+	}
+	defer reader.Close()
+
+	// Start reading postgres statements.
+	ps := newPostgreStream(ctx, reader, maxRowSize, &unsupportedStmtLogger{} /* unsupportedStmtLogger */)
+	for {
+		stmt, err := ps.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return handler, errors.Wrap(err, "postgres parse error")
+		}
+		if err := bufferDDLPostgresStatement(ctx, evalCtx, stmt, p, parentID, &handler); err != nil {
+			return handler, err
+		}
+	}
+	return handler, nil
+}
+
+func runDDLStatementsFromDumpFile(
+	ctx context.Context, p sql.JobExecContext, h postgresDDLHandler,
+) error {
+	for _, stmt := range h.bufferedDDLStmts {
+		_, err := p.ExecCfg().InternalExecutor.ExecEx(ctx, "import-pgdump-ddl", nil, /* txn */
```

Comment:
So just to clarify, this will run each statement in a new, separate session and transaction for now, but the hope is internalExec improvements allow switching this in the future?